### PR TITLE
Jacobian diagnostic

### LINKED ISSUE #4586: Domain diagnostic: jacobian
# Feature request:

### Component:

- [x] Code
- [ ] Documentation
- [ ] Build system
- [ ] Continuous integration

### Desired feature:

- [ ] Add a diagnostic that compares the jacobian provided by the map to just numerically computing the jacobian.

### PR DESCRIPTION
**Pull request by geoffrey4444**
## Proposed changes

Implements a diagnostic that compares the analytic and numeric Jacobians. Resolves #4586.

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `new feature` if appropriate.

### CODE REVIEW

File: src/Domain/JacobianDiagnostic.hpp
Comment (nilsvu) [inline src/Domain/JacobianDiagnostic.hpp]: I'd say take the derivative in the `jacobian_diagnostic` function so calling the function is agnostic to the transpose stuff. Just pass the inertial coords and the mesh to `jacobian_diagnostic`.

File: src/Domain/JacobianDiagnostic.hpp
Comment (nilsvu) [inline src/Domain/JacobianDiagnostic.hpp]: I don't remember which map compositions we store in the DataBox in evolutions, but I know we have both the grid and the inertial inv jacobian in there. So it seems easiest to take the `domain::Tags::InverseJacobian<Dim, Frame::ElementLogical, TargetFrame` as an argument directly instead of the map.

File: src/Domain/JacobianDiagnostic.cpp
Comment (nilsvu) [inline src/Domain/JacobianDiagnostic.cpp]: delete iwyu pragmas

File: src/Domain/JacobianDiagnostic.cpp
Comment (nilsvu) [inline src/Domain/JacobianDiagnostic.cpp]: Use `equal_within_roundoff` instead of `eps_for_ratio`.

File: src/Domain/JacobianDiagnostic.cpp
Comment (nilsvu) [inline src/Domain/JacobianDiagnostic.cpp]: Use `equal_within_roundoff` to divide only if both numerator and denominator are nonzero.

File: src/Domain/JacobianDiagnostic.cpp
Comment (nilsvu) [inline src/Domain/JacobianDiagnostic.cpp]: Minor suggestions: I was confused these are named `difference` but no difference is computed afaik. Maybe name them `analytic_jacobian_sum` and `numeric_jacobian_sum`? You can also simplify this code like this if you want:

```cpp
const size_t num_points = analytic_jacobian.begin()->size();
tnsr::i<DataVector, Dim, Frame::ElementLogical> analytic_jacobian_sum{num_points, 0.};
```

(Haven't tested this, but I think `Tensor` has a constructor that's forwarded to all components. You only need `make_with_value` if you're working generically with both DataVector and double)

**Review by nilsvu — CHANGES_REQUESTED** [review]

File: src/Domain/JacobianDiagnostic.hpp
Comment (nilsvu) [inline src/Domain/JacobianDiagnostic.hpp]: Let's return a vector here to look at the logical directions independently

File: src/Domain/JacobianDiagnostic.hpp
Comment (geoffrey4444) [reply src/Domain/JacobianDiagnostic.hpp]: I'd like to leave this as-is so that the free function only takes in tensors, making it easy to test

Comment (nilsvu) [reply src/Domain/JacobianDiagnostic.hpp]: This won't work great with the python bindings because we have to compute the logical derivative outside the function. If the function only takes jacobian, coordinates, and mesh then it works with the python bindings out of the box. Suggestion: keep the current function that takes only tensors, name it `jacobian_diagnostic_impl`, and test it the way you currently do. Move the function that's currently in the compute item to a free function and name it `jacobian_diagnostic`, so we can call it easily from python. wdyt?

**Comment by geoffrey4444** [comment]
> This won't work great with the python bindings because we have to compute the logical derivative outside the function. If the function only takes jacobian, coordinates, and mesh then it works with the python bindings out of the box. Suggestion: keep the current function that takes only tensors, name it `jacobian_diagnostic_impl`, and test it the way you currently do. Move the function that's currently in the compute item to a free function and name it `jacobian_diagnostic`, so we can call it easily from python. wdyt? 

I think that could work, but what if we name both `jacobian_diagnostic`...since they take different parameters, why not use the same name?

**Comment by nilsvu** [comment]
sure that sounds fine

File: src/Domain/JacobianDiagnostic.hpp
Comment (geoffrey4444) [reply src/Domain/JacobianDiagnostic.hpp]: As we discussed below, I'll do this except name both `jacobian_diagnostic`

File: src/Domain/JacobianDiagnostic.cpp
Comment (geoffrey4444) [reply src/Domain/JacobianDiagnostic.cpp]: This check is no longer necessary, because instead I changed this so instead of summing the components, it does sqrt(sum(square))) of the components, so that you don't get zero unless all components of the jacobian are zero. So I no longer do this check at all.

File: src/Domain/JacobianDiagnostic.cpp
Comment (geoffrey4444) [reply src/Domain/JacobianDiagnostic.cpp:51]: This check is no longer necessary, because instead I changed this so instead of summing the components, it does sqrt(sum(square))) of the components, so that you don't get zero unless all components of the jacobian are zero. So I no longer do this check at all.

Comment (geoffrey4444) [reply src/Domain/JacobianDiagnostic.cpp]: I don't want to ignore a term if the denominator is zero; instead, I want to get a huge number (so that I can see in the diagnostic that something is really wrong). So I think I should leave the epsilon here.

File: src/Domain/JacobianDiagnostic.cpp
Comment (nilsdeppe) [inline src/Domain/JacobianDiagnostic.cpp]: Why not just take the jacobian as an argument instead of the inverse jacobian?

File: src/Domain/JacobianDiagnostic.cpp
Comment (nilsdeppe) [inline src/Domain/JacobianDiagnostic.cpp:31]: You can do this with 1 `Scalar<DataVector>` as temporary storage and just use `abs` instead of RMS.

**Review by nilsdeppe — CHANGES_REQUESTED** [review]

File: src/Domain/JacobianDiagnostic.cpp
Comment (nilsvu) [inline src/Domain/JacobianDiagnostic.cpp:27]: (nitpick, optional) move up one line and use in `destructive_resize`

File: src/Domain/JacobianDiagnostic.hpp
Comment (nilsvu) [inline src/Domain/JacobianDiagnostic.hpp]: I think you can delete `MappedCoordsTag` and instead write `domain::Tags::Coordinates<Dim, TargetFrame>`.

Please update the docs of struct struct as well.

File: src/Domain/JacobianDiagnostic.cpp
Comment (nilsvu) [inline src/Domain/JacobianDiagnostic.cpp]: Maybe instead of the epsilon thing you just set the quantity to NaN (or `quiet_NaN`) if the denominator is zero? That should propagate fine to ParaView.

File: src/Domain/JacobianDiagnostic.cpp
Comment (nilsvu) [inline src/Domain/JacobianDiagnostic.cpp]: I think you can delete these two lines

File: src/Domain/JacobianDiagnostic.hpp
Comment (nilsvu) [inline src/Domain/JacobianDiagnostic.hpp:85]: delete `typename`

File: tests/Unit/Domain/Python/Test_JacobianDiagnostic.py
Comment (nilsvu) [inline tests/Unit/Domain/Python/Test_JacobianDiagnostic.py]: `analytic_jacobian**2`

**Review by nilsvu — CHANGES_REQUESTED** [review]
Please squash

**Comment by nilsvu** [comment]
@geoffrey4444 would you be willing to add a pybinding for this as well? Can also be a follow-up PR.

**Comment by geoffrey4444** [comment]
> @geoffrey4444 would you be willing to add a pybinding for this as well? Can also be a follow-up PR.

I do want to add pybindings, but in a follow-up PR

File: src/Domain/JacobianDiagnostic.cpp
Comment (geoffrey4444) [reply src/Domain/JacobianDiagnostic.cpp]: I prefer leaving the epsilon as-is

Comment (nilsdeppe) [reply src/Domain/JacobianDiagnostic.cpp]: You can't get a zero denominator anymore. That was only because the different terms could have canceling signs, but with absolute/RMS the denominator has to be positive, otherwise the Jacobian isn't invertible and the map is definitely bad.

File: tests/Unit/Domain/Python/Test_JacobianDiagnostic.py
Comment (geoffrey4444) [reply tests/Unit/Domain/Python/Test_JacobianDiagnostic.py]: I'm no longer using RMS, but I'll keep in mind that **2 is ok to use in future python tests

File: src/Domain/JacobianDiagnostic.cpp
Comment (nilsvu) [reply src/Domain/JacobianDiagnostic.cpp]: so, delete the eps?

**Comment by geoffrey4444** [comment]
Thanks for the reviews @nilsvu and @nilsdeppe ! I responded stop your remaining comments and squashed...this is ready for you to have another look when you get a chance.

File: src/Domain/JacobianDiagnostic.cpp
Comment (nilsdeppe) [inline src/Domain/JacobianDiagnostic.cpp]: I think since will almost certainly use this we want to optimize a bit further:
```
jacobian_diag->get(i_hat) = -abs(abs(analytic_jacobian.get(0, i_hat));
for (size_t i = 1; ...) { ... }
get(jacobian_component_sum) =  abs(numeric_jacobian_transpose.get(i_hat, 0));
for (size_t i = 1; ...) { ... }
```
and then we can also eliminate the `0.0` on line 28, on 37  and on 44.

**Review by nilsdeppe — COMMENTED** [review]
A small pedantic performance consideration since I think we'll use this in AMR and so we will want this to be quite fast

Edit: you can squash immediately!

**Review by nilsvu — DISMISSED** [review]
don't wait for my re-approval

File: src/Domain/JacobianDiagnostic.cpp
Comment (geoffrey4444) [reply src/Domain/JacobianDiagnostic.cpp]: Done, squashed, and force-pushed...good suggestion!

File: src/Domain/JacobianDiagnostic.cpp
Comment (nilsdeppe) [inline src/Domain/JacobianDiagnostic.cpp]: Uh oh, memory allocation in a loop! Can you go back to allocating outside the loop. Just move this line to line 30, nothing else needs to be changed!

Comment (geoffrey4444) [reply src/Domain/JacobianDiagnostic.cpp]: 🤦‍♂️that's what I get for trying to get it in 5 minutes before lunch! "O, let's declare this when we need it!", I thought.

Anyway, thanks for catching, fixed!

p.s. I saw some CI failures, but they looked not related to my PR but to python

**Review by nilsdeppe — DISMISSED** [review]

**Comment by nilsdeppe** [comment]
You have missing linker dependencies. I think you need `Domain` to depend on `Spectral` as a `PUBLIC` instead of `INTERFACE` library. I.e. move `Spectral` from `INTERFACE` to `PUBLIC` in `src/Domain/CMakeLists.txt`

File: src/Domain/JacobianDiagnostic.hpp
Comment (nilsvu) [inline src/Domain/JacobianDiagnostic.hpp]: <s>No it's not `Spectral`.</s> Move this include to the cpp, then depend on `PRIVATE LinearOperators`. Edit: Actually yes, depend on `PUBLIC Spectral` also and include Mesh.hpp here`.

Also delete the DeterminantAndInverse include.

File: src/Domain/JacobianDiagnostic.hpp
Comment (nilsvu) [inline src/Domain/JacobianDiagnostic.hpp:22]: Typo "diagnostic", but you can fix it some other time

**Review by nilsvu — APPROVED** [review]

**Pull request merged**
